src/metabase/server/middleware/browser_cookie.clj

```diff
@@ -0,0 +1,36 @@
+(ns metabase.server.middleware.browser-cookie
+  "Middleware that sets a permanent browser identifier cookie so we can identify logins from new browsers. This is
+  mostly so we can send people 'login from a new device' emails the first time they log in with a new browser. If this
+  cookie is deleted, it's fine; the user will just get an email saying they logged in from a new device next time
+  they log in."
+  (:require [java-time :as t]
+            [metabase.util.schema :as su]
+            [ring.util.response :as resp]
+            [schema.core :as s])
+  (:import java.util.UUID))
+
+(def ^:private browser-id-cookie-name "metabase.DEVICE")
+
+;; This cookie doesn't need to be secure, because it's only used for notification purposes
+(def ^:private cookie-options
+  {:http-only true
+   :path      "/"
+   :same-site :lax
+   ;; Set the cookie to expire 20 years from now. That should be sufficient
+   :expires   (t/format :rfc-1123-date-time (t/plus (t/zoned-date-time) (t/years 20)))})
+
+(s/defn ^:private add-browser-id-cookie [response browser-id :- su/NonBlankString]
+  (resp/set-cookie response browser-id-cookie-name browser-id cookie-options))
+
+(defn ensure-browser-id-cookie
+  "Set a permanent browser identifier cookie if one is not already set."
+  [handler]
+  (fn [request respond raise]
+    (if-let [browser-id (get-in request [:cookies browser-id-cookie-name :value])]
+      (handler (assoc request :browser-id browser-id) respond raise)
+      (let [browser-id (str (UUID/randomUUID))]
```

Comment:
we just check that this key is set. Should we just use a static string so it doesn't even have the appearance of tracking? We don't connect this UUID to anything so its really just a boolean.

Reply:
We do use it. There's a `device_id` column in the new LoginHistory table. That's how we know if a given user has logged in from a given device before. If it's just a boolean, we wouldn't be able to tell if it was a new login for you if someone did it from a shared computer. It would also be trivial for attackers to make their own browser ID cookie and bypass the login from new device notification

Reply:
oh i see. i was just thinking if we set the cookie for the first time we did an action and it didn't matter what the device_id is.